# Project : WeRateDog's Twitter Account Data Investigation  

### Introduction 
<a name="top"></a>
> The WeRateDog's Twitter Archive Enhanced contained over 5000 twitter data on dog rating. The goal for this project is to gather, access and clean the data into tidy and high quality data table. 

### Table of Content
<ul>
    <li><a href="#dg">Data Gathering</a></li>
    <li><a href="#ad">Assessing Data</a></li>
    <li><a href="#cd">Clean Data</a></li>
    <li><a href="#sd">Store Data</a></li>
</ul>

<a id='dg'></a>
### Data Gathering 
<a href="#top">Back to Table of Content</a>

In [26]:
#importing libraries
import pandas as pd
import numpy as np
import requests
import os
import json
import time
import tweepy

### 1. Twitter Archive Enhanced 
> The file contain over 2000 rows of dog rating and data on twitter such as tweet id, source, name and rating number. 

In [64]:
df_1 = pd.read_csv('twitter-archive-enhanced.csv')

In [65]:
df_1.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [66]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

 ### 2. Tweet image predictions
 > The tweet image predictions contain what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network. This file (image_predictions.tsv) is hosted on Udacity's servers and should be downloaded programmatically using the Requests library and the following URL: https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv

In [67]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [68]:
file_images = url.split('/')[-1]

with open(os.path.join('./', file_images), mode = 'wb') as file:
    file.write(response.content)

In [71]:
df_2 = pd.read_csv(file_images, sep = '\\t', engine = 'python')
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [72]:
df_2.head(10)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5  666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6  666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7  666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8  666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9  666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   
5        1    Bernese_mountain_dog  0.651137    True    English_springer   
6        1              box_turtle  0.933012   False          mud_turtle   
7        1                    chow  0.692517    True     Tibetan_mastiff   
8        1           shopping_cart  0.962465   False     shopping_basket   
9        1        miniature_poodle  0.201493    True            komondor   

    p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0  0.156665    True            Shetland_sheepdog  0.061428    True  
1  0.074192    True          Rhodesian_ridgeback  0.072010    True  
2  0.138584    True                   bloodhound  0.116197    True  
3  0.360687    True           miniature_pinscher  0.222752    True  
4  0.243682    True                     Doberman  0.154629    True  
5  0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6  0.045885   False                     terrapin  0.017885   False  
7  0.058279    True                     fur_coat  0.054449   False  
8  0.014594   False             golden_retriever  0.007959    True  
9  0.192305    True  soft-coated_wheaten_terrier  0.082086    True

### 3. Tweet Json Data
> Using the tweet IDs in the WeRateDogs Twitter archive, query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file. <strong>For some reason, twitter has never get back to me for the developer account after I applied, I will be using the file that is already prepared</strong>. 

In [73]:
df_3 = pd.read_json('tweet-json.txt', lines=True)

In [74]:
df_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2354 non-null   datetime64[ns, UTC]
 1   id                             2354 non-null   int64              
 2   id_str                         2354 non-null   int64              
 3   full_text                      2354 non-null   object             
 4   truncated                      2354 non-null   bool               
 5   display_text_range             2354 non-null   object             
 6   entities                       2354 non-null   object             
 7   extended_entities              2073 non-null   object             
 8   source                         2354 non-null   object             
 9   in_reply_to_status_id          78 non-null     float64            
 10  in_reply_to_status_id_st

In [75]:
df_3.shape

(2354, 31)

<a id='ad'></a>
### Assess Data
<a href="#top">Back to Table of Content</a>

> <strong>Twitter Archive Enhanced</strong>

In [77]:
df_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

> From the information above I can summarize the quality and tidiness issues as followed:
> - Only 181 retweets (retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)
> - Only 78 replies (in_reply_to_status_id, in_reply_to_user_id)
> - 2297 tweets with expanded_urls indicating 59 tweets with missing data
> - The timestamp field is in string format (object)
> - There are 4 columns for dog stages (doggo, floofer, pupper, puppo)
> - The columns related to retweets are not applicable for original tweets
> - The columns related to replies are not applicable for original tweets
> - The timestamp column has dates in string form.
> - Some of the rows have invalid strings in the name column, e.g. "a", "an", "in". These words are all the 3rd word in the tweet.
> - Values of "None" in the name column.
> - Tidiness issues were that the numerical data are located in the far right while long string are on the left. 

> 1.Data quality: <strong>name</strong>

In [80]:
#checking name for any odd value
df_1.name.value_counts().head(20)

None       745
a           55
Charlie     12
Lucy        11
Cooper      11
Oliver      11
Penny       10
Tucker      10
Lola        10
Bo           9
Winston      9
the          8
Sadie        8
Toby         7
Buddy        7
Daisy        7
an           7
Bailey       7
Milo         6
Jax          6
Name: name, dtype: int64

> Names are capitalize while name with "a" are not a name. Lets find out.

In [82]:
err_names = df_1.name.str.contains('^[a-z]', regex = True)
df_1[err_names].name.value_counts().sort_index()

a               55
actually         2
all              1
an               7
by               1
getting          2
his              1
incredibly       1
infuriating      1
just             4
life             1
light            1
mad              2
my               1
not              2
officially       1
old              1
one              4
quite            4
space            1
such             1
the              8
this             1
unacceptable     1
very             5
Name: name, dtype: int64

In [83]:
(len(df_1[err_names]))

109

> 109 valid name 

> 2. Data quality: <strong>expanded_urls</strong>

In [84]:
# check the number of tweets with missing urls
sum(df_1.expanded_urls.isna())

59

In [88]:
# check how many tweets with missing urls are replies or retweets?
sum(df_1.expanded_urls.isna() & \
    (df_1.in_reply_to_status_id.notnull() | \
     df_1.retweeted_status_id.notnull()))

56

In [91]:
# find out tweets that are missing url, not replies and not retweet
mask = df_1.expanded_urls.isna() & \
       df_1.in_reply_to_status_id.isna() & \
       df_1.retweeted_status_id.isna()
df_1[mask][['tweet_id', 'text', 'expanded_urls']]

tweet_id                                               text  \
375   828361771580813312  Beebop and Doobert should start a band 12/10 w...   
707   785515384317313025     Today, 10/10, should be National Dog Rates Day   
1445  696518437233913856    Oh my god 10/10 for every little hot dog pupper   

     expanded_urls  
375            NaN  
707            NaN  
1445           NaN

> 3 tweets with NAN in expanded_urls

3. Data quality: <strong>rating_denominator</strong>

In [92]:
#checking odds denomator rating number and find out how many of them 
df_1.rating_denominator.value_counts().sort_index(ascending = False)

170       1
150       1
130       1
120       1
110       1
90        1
80        2
70        1
50        3
40        1
20        2
16        1
15        1
11        3
10     2333
7         1
2         1
0         1
Name: rating_denominator, dtype: int64

In [93]:
sum(df_1.rating_denominator != 10)

23

In [95]:
#find out what tweet has a denominator of more than 10m not a reply and not retweet
deno_find = (df_1.rating_denominator != 10) & \
            df_1.in_reply_to_status_id.isna() & \
            df_1.retweeted_status_id.isna()

deno_cols = ['text', 'tweet_id', 'rating_numerator', 'rating_denominator']
df_1[deno_find][deno_cols]

text            tweet_id  \
433   The floofs have been released I repeat the flo...  820690176645140481   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...  810984652412424192   
902   Why does this never happen at my front door......  758467244762497024   
1068  After so many requests, this is Bretagne. She ...  740373189193256964   
1120  Say hello to this unbelievably well behaved sq...  731156023742988288   
1165  Happy 4/20 from the squad! 13/10 for all https...  722974582966214656   
1202  This is Bluebert. He just saw that both #Final...  716439118184652801   
1228  Happy Saturday here's 9 puppers on a bench. 99...  713900603437621249   
1254  Here's a brigade of puppers. All look very pre...  710658690886586372   
1274  From left to right:\nCletus, Jerome, Alejandro...  709198395643068416   
1351  Here is a whole flock of puppers.  60/50 I'll ...  704054845121142784   
1433  Happy Wednesday here's a bucket of pups. 44/40...  697463031882764288   
1635  Someone help the girl is being mugged. Several...  684222868335505415   
1662  This is Darrel. He just robbed a 7/11 and is i...  682962037429899265   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...  677716515794329600   
1843  Here we have an entire platoon of puppers. Tot...  675853064436391936   
2335  This is an Albanian 3 1/2 legged  Episcopalian...  666287406224695296   

      rating_numerator  rating_denominator  
433                 84                  70  
516                 24                   7  
902                165                 150  
1068                 9                  11  
1120               204                 170  
1165                 4                  20  
1202                50                  50  
1228                99                  90  
1254                80                  80  
1274                45                  50  
1351                60                  50  
1433                44                  40  
1635               121                 110  
1662                 7                  11  
1779               144                 120  
1843                88                  80  
2335                 1                   2

In [96]:
sum(deno_find)

17

> Here are 17 tweets that have denominator of more than 10. Reading from the text we can assume that the photo have more than one puppies, the obvious one are at 1274.

4. Data quality: <strong>rating_numerator</strong>

In [97]:
#checking odds numerator rating number and find out how many of them 
df_1.rating_numerator.value_counts().sort_index(ascending = False)

1776      1
960       1
666       1
420       2
204       1
182       1
165       1
144       1
143       1
121       1
99        1
88        1
84        1
80        1
75        2
60        1
50        1
45        1
44        1
27        1
26        1
24        1
20        1
17        1
15        2
14       54
13      351
12      558
11      464
10      461
9       158
8       102
7        55
6        32
5        37
4        17
3        19
2         9
1         9
0         2
Name: rating_numerator, dtype: int64

In [105]:
sum(df_1.rating_numerator >= 15)

28

In [107]:
#lets look at tweets that has the denominator of 10 
denom_10 = df_1[df_1.rating_denominator == 10]
denom_10.rating_numerator.value_counts().sort_index(ascending = False)

1776      1
666       1
420       2
182       1
75        2
27        1
26        1
17        1
15        2
14       54
13      351
12      558
11      463
10      461
9       156
8       102
7        54
6        32
5        37
4        15
3        19
2         9
1         8
0         2
Name: rating_numerator, dtype: int64

In [108]:
sum(denom_10.rating_numerator >= 15)

12

> There are 12 data with numerator of >= 15 and denominator of 10. 

In [111]:
#Lets find out the tweets of numerator of >=15 which have 10 as denomiator, not reply and retweet.
deno_10_nume_15 = (df_1.rating_denominator == 10) & \
            (df_1.rating_numerator >= 15) & \
            df_1.in_reply_to_status_id.isna() & \
            df_1.retweeted_status_id.isna()
deno_10_nume_15_cols = ['text', 'tweet_id', 'rating_numerator', 'rating_denominator']
df_1[deno_10_nume_15][deno_10_nume_15_cols]

text            tweet_id  \
695   This is Logan, the Chow who lived. He solemnly...  786709082849828864   
763   This is Sophie. She's a Jubilant Bush Pupper. ...  778027034220126208   
979   This is Atticus. He's quite simply America af....  749981277374128128   
1712  Here we have uncovered an entire battalion of ...  680494726643068929   
2074  After so many requests... here you go.\n\nGood...  670842764863651840   

      rating_numerator  rating_denominator  
695                 75                  10  
763                 27                  10  
979               1776                  10  
1712                26                  10  
2074               420                  10

5. Data quality: <strong>Source</strong>

In [115]:
df_1.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

> There are 4 sources. These can be display by using string before the "</ a>"

6. Data quality: <strong>doggo, floofer, pupper, and puppo</strong>

In [117]:
df_1.doggo.value_counts()

None     2259
doggo      97
Name: doggo, dtype: int64

In [119]:
df_1.floofer.value_counts()

None       2346
floofer      10
Name: floofer, dtype: int64

In [120]:
df_1.pupper.value_counts()

None      2099
pupper     257
Name: pupper, dtype: int64

In [121]:
df_1.puppo.value_counts()

None     2326
puppo      30
Name: puppo, dtype: int64

> Assuming from the none value, it might be that each word spelled in different way. Lets find out!

In [122]:
print(df_1[df_1.floofer == 'floofer'].text)
print(df_1[df_1.doggo == 'doggo'].text)
print(df_1[df_1.pupper == 'pupper'].text)
print(df_1[df_1.puppo == 'puppo'].text)

46      Meet Grizzwald. He may be the floofiest floofe...
200     At first I thought this was a shy doggo, but i...
582     This is Doc. He takes time out of every day to...
774     Atlas rolled around in some chalk and now he's...
984     This is Blu. He's a wild bush Floofer. I wish ...
1022    Here's a golden floofer helping with the groce...
1091    Just wanted to share this super rare Rainbow F...
1110    This is Moose. He's a Polynesian Floofer. Dapp...
1534    Here we are witnessing a rare High Stepping Al...
1614    Say hello to Petrick. He's an Altostratus Floo...
Name: text, dtype: object
9       This is Cassie. She is a college pup. Studying...
43      Meet Yogi. He doesn't have any important dog m...
99      Here's a very large dog. He has a date later. ...
108     This is Napolean. He's a Raggedy East Nicaragu...
110     Never doubt a doggo 14/10 https://t.co/AbBLh2FZCH
                              ...                        
1117    This is Kyle (pronounced 'Mitch'). He 

> Take a look at index 45, the tweet has different spelling for floofer which can also be floofiest. 

> <strong>Tweet image predictions</strong>

In [113]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [123]:
df_2.head(10)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5  666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6  666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7  666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8  666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9  666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   
5        1    Bernese_mountain_dog  0.651137    True    English_springer   
6        1              box_turtle  0.933012   False          mud_turtle   
7        1                    chow  0.692517    True     Tibetan_mastiff   
8        1           shopping_cart  0.962465   False     shopping_basket   
9        1        miniature_poodle  0.201493    True            komondor   

    p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0  0.156665    True            Shetland_sheepdog  0.061428    True  
1  0.074192    True          Rhodesian_ridgeback  0.072010    True  
2  0.138584    True                   bloodhound  0.116197    True  
3  0.360687    True           miniature_pinscher  0.222752    True  
4  0.243682    True                     Doberman  0.154629    True  
5  0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6  0.045885   False                     terrapin  0.017885   False  
7  0.058279    True                     fur_coat  0.054449   False  
8  0.014594   False             golden_retriever  0.007959    True  
9  0.192305    True  soft-coated_wheaten_terrier  0.082086    True

In [125]:
df_2.tail(10)

tweet_id                                          jpg_url  \
2065  890240255349198849  https://pbs.twimg.com/media/DFrEyVuW0AAO3t9.jpg   
2066  890609185150312448  https://pbs.twimg.com/media/DFwUU__XcAEpyXI.jpg   
2067  890729181411237888  https://pbs.twimg.com/media/DFyBahAVwAAhUTd.jpg   
2068  890971913173991426  https://pbs.twimg.com/media/DF1eOmZXUAALUcq.jpg   
2069  891087950875897856  https://pbs.twimg.com/media/DF3HwyEWsAABqE6.jpg   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                        p1   p1_conf  p1_dog                  p2  \
2065        1                  Pembroke  0.511319    True            Cardigan   
2066        1             Irish_terrier  0.487574    True        Irish_setter   
2067        2                Pomeranian  0.566142    True          Eskimo_dog   
2068        1               Appenzeller  0.341703    True       Border_collie   
2069        1  Chesapeake_Bay_retriever  0.425595    True       Irish_terrier   
2070        2                    basset  0.555712    True    English_springer   
2071        1               paper_towel  0.170278   False  Labrador_retriever   
2072        1                 Chihuahua  0.716012    True            malamute   
2073        1                 Chihuahua  0.323581    True            Pekinese   
2074        1                    orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
2065  0.451038    True                    Chihuahua  0.029248    True  
2066  0.193054    True     Chesapeake_Bay_retriever  0.118184    True  
2067  0.178406    True                     Pembroke  0.076507    True  
2068  0.199287    True                    ice_lolly  0.193548   False  
2069  0.116317    True              Indian_elephant  0.076902   False  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False

In [127]:
df_2.p1_dog.value_counts()

True     1532
False     543
Name: p1_dog, dtype: int64

In [128]:
df_2.p2_dog.value_counts()

True     1553
False     522
Name: p2_dog, dtype: int64

In [126]:
df_2.p3_dog.value_counts()

True     1499
False     576
Name: p3_dog, dtype: int64

> From the investigation there are multiples entries that the algorithm failed to predict dog. Let's find out what tweet id that didn't perdict dog.

In [131]:
#checking 10 with unidentify dog breed
df_2[~df_2.p1_dog & \
     ~df_2.p2_dog & \
     ~df_2.p3_dog][['tweet_id', 'p1', 'p1_dog','p2', 'p2_dog', 'p3', 'p3_dog']].head(10)

tweet_id                p1  p1_dog                   p2  p2_dog  \
6   666051853826850816        box_turtle   False           mud_turtle   False   
17  666104133288665088               hen   False                 cock   False   
18  666268910803644416  desktop_computer   False                 desk   False   
21  666293911632134144  three-toed_sloth   False                otter   False   
25  666362758909284353        guinea_pig   False                skunk   False   
29  666411507551481857              coho   False           barracouta   False   
45  666786068205871104             snail   False                 slug   False   
50  666837028449972224       triceratops   False            armadillo   False   
51  666983947667116034              swab   False            chain_saw   False   
53  667012601033924608             hyena   False  African_hunting_dog   False   

                p3  p3_dog  
6         terrapin   False  
17       partridge   False  
18        bookcase   False  
21  great_grey_owl   False  
25         hamster   False  
29             gar   False  
45           acorn   False  
50   common_iguana   False  
51             wig   False  
53          coyote   False

In [136]:
#check how many dog that are not identify in those three perdiction
sum(~df_2.p1_dog & \
    ~df_2.p2_dog & \
    ~df_2.p3_dog)

324

> <strong>Tweet Json Data</strong>

> As suggesting from udacity, I will drop all columns <strong>except id, retweet_count and favorite_count</strong>. In this data assessing, I will also neglect other columns except those three. 

In [137]:
df_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2354 non-null   datetime64[ns, UTC]
 1   id                             2354 non-null   int64              
 2   id_str                         2354 non-null   int64              
 3   full_text                      2354 non-null   object             
 4   truncated                      2354 non-null   bool               
 5   display_text_range             2354 non-null   object             
 6   entities                       2354 non-null   object             
 7   extended_entities              2073 non-null   object             
 8   source                         2354 non-null   object             
 9   in_reply_to_status_id          78 non-null     float64            
 10  in_reply_to_status_id_st

In [148]:
df_3['id'].describe()

count    2.354000e+03
mean     7.426978e+17
std      6.852812e+16
min      6.660209e+17
25%      6.783975e+17
50%      7.194596e+17
75%      7.993058e+17
max      8.924206e+17
Name: id, dtype: float64

In [149]:
df_3['retweet_count'].describe()

count     2354.000000
mean      3164.797366
std       5284.770364
min          0.000000
25%        624.500000
50%       1473.500000
75%       3652.000000
max      79515.000000
Name: retweet_count, dtype: float64

In [150]:
df_3['favorite_count'].describe()

count      2354.000000
mean       8080.968564
std       11814.771334
min           0.000000
25%        1415.000000
50%        3603.500000
75%       10122.250000
max      132810.000000
Name: favorite_count, dtype: float64

> I will combine this data with df_1: Tweet Archive Enhanced table

<a id='cd'></a>
### Clean Data
<a href="#top">Back to Table of Content</a>

In [251]:
# I will create a copies of the dataframe for cleaning 
df_1_clean = df_1.copy()
df_2_clean = df_2.copy()
df_3_clean = df_3.copy()

<strong>Tweet Archive Enhanced</strong> aka. <strong>df_1</strong> data quality and tidiness issues:
- 1. Drop all 181 retweets (retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)
- 2. Drop all 78 replies (in_reply_to_status_id, in_reply_to_user_id)
- 3. Drop 59 tweets with missing data in extended_urls columns 
- 4. The timestamp field is in string format (object)
- 5. Drop all 109 tweets that are invalid name 
- 6. Sources can be extracting from the "a" HTML tag
- 7. Drop rating denominator not equal to 10
- 8. Drop rating numerator that are greater or equal from 15  
- 9. There are 4 columns for dog stages (doggo, floofer, pupper, puppo) 

<strong>Tweet Image Perdiction</strong> aka. <strong>df_2 data</strong> quality issues: 
- From the investigation there are multiples entires that the algorithm failed to perdict dog.

<strong>Json Tweet Data</strong> aka. <strong>df_3</strong> data tidiness issue: 
- From the investigation, I will combine the data to df_1 (Tweet Archive Enhanced)

<strong>Tweet Archive Enhanced</strong> aka. <strong>df_1</strong>

> DATA QUALITY issues: 

##### Define

1. Drop all rows with 181 retweets (retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp)

##### Code

In [252]:
df_1_clean = df_1_clean[df_1_clean.retweeted_status_id.isna()]

##### Test

In [253]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2175 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2175 non-null   object 
 4   source                      2175 non-null   object 
 5   text                        2175 non-null   object 
 6   retweeted_status_id         0 non-null      float64
 7   retweeted_status_user_id    0 non-null      float64
 8   retweeted_status_timestamp  0 non-null      object 
 9   expanded_urls               2117 non-null   object 
 10  rating_numerator            2175 non-null   int64  
 11  rating_denominator          2175 non-null   int64  
 12  name                        2175 non-null   object 
 13  doggo                       2175 

##### Define 

2. Drop all 78 replies (in_reply_to_status_id, in_reply_to_user_id)

##### Code

In [254]:
df_1_clean = df_1_clean[df_1_clean.in_reply_to_status_id.isna()]

##### Test

In [255]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2097 non-null   int64  
 1   in_reply_to_status_id       0 non-null      float64
 2   in_reply_to_user_id         0 non-null      float64
 3   timestamp                   2097 non-null   object 
 4   source                      2097 non-null   object 
 5   text                        2097 non-null   object 
 6   retweeted_status_id         0 non-null      float64
 7   retweeted_status_user_id    0 non-null      float64
 8   retweeted_status_timestamp  0 non-null      object 
 9   expanded_urls               2094 non-null   object 
 10  rating_numerator            2097 non-null   int64  
 11  rating_denominator          2097 non-null   int64  
 12  name                        2097 non-null   object 
 13  doggo                       2097 

##### Define

3. Drop all 59 tweets with missing data in expanded_urls

##### Code

In [256]:
df_1_clean = df_1_clean[df_1_clean.expanded_urls.notnull()]

##### Test

In [257]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2094 non-null   int64  
 1   in_reply_to_status_id       0 non-null      float64
 2   in_reply_to_user_id         0 non-null      float64
 3   timestamp                   2094 non-null   object 
 4   source                      2094 non-null   object 
 5   text                        2094 non-null   object 
 6   retweeted_status_id         0 non-null      float64
 7   retweeted_status_user_id    0 non-null      float64
 8   retweeted_status_timestamp  0 non-null      object 
 9   expanded_urls               2094 non-null   object 
 10  rating_numerator            2094 non-null   int64  
 11  rating_denominator          2094 non-null   int64  
 12  name                        2094 non-null   object 
 13  doggo                       2094 

##### Define

4. The timestamp field is in string format (object) should be datetime datatype

##### Code

In [258]:
df_1_clean['timestamp'] = pd.to_datetime(df_1_clean.timestamp)
df_1_clean['timestamp'] = df_1_clean.timestamp.dt.floor('s')

##### Test

In [259]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2094 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2094 non-null   datetime64[ns, UTC]
 4   source                      2094 non-null   object             
 5   text                        2094 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      object             
 9   expanded_urls               2094 non-null   object             
 10  rating_numerator            2094 non-null   int64           

##### Define

5. 109 tweets that are invalid name in name columns all of them are in lower case which will replace with "none" 

##### Code

In [260]:
names_mask = df_1_clean.name.str.contains('^[a-z]', regex = True)
df_1_clean[names_mask].name.value_counts().sort_index()

a               55
actually         2
all              1
an               6
by               1
getting          2
his              1
incredibly       1
infuriating      1
just             3
life             1
light            1
mad              1
my               1
not              2
officially       1
old              1
one              4
quite            3
space            1
such             1
the              8
this             1
unacceptable     1
very             4
Name: name, dtype: int64

In [261]:
df_1_clean.loc[names_mask, 'name'] = "none"

##### Test

In [262]:
df_1_clean[names_mask].name.value_counts().sort_index()

none    104
Name: name, dtype: int64

In [263]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2094 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2094 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2094 non-null   datetime64[ns, UTC]
 4   source                      2094 non-null   object             
 5   text                        2094 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      object             
 9   expanded_urls               2094 non-null   object             
 10  rating_numerator            2094 non-null   int64           

##### Define

6. The sources can be extreacting from the string after the "a" HTML tag 

##### Code

In [264]:
df_1_clean['source'] = df_1_clean['source'].str.extract('^<a.+>(.+)</a>$')

##### Test

In [265]:
df_1_clean.source.value_counts()

Twitter for iPhone     1962
Vine - Make a Scene      91
Twitter Web Client       30
TweetDeck                11
Name: source, dtype: int64

##### Define

7. Drop the values with the denominator not equal to 10 

##### Code

In [266]:
df_1_clean = df_1_clean[df_1_clean.rating_denominator == 10]

##### Test

In [267]:
df_1_clean.rating_denominator.value_counts().sort_index(ascending = False)

10    2077
Name: rating_denominator, dtype: int64

In [268]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2077 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2077 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2077 non-null   datetime64[ns, UTC]
 4   source                      2077 non-null   object             
 5   text                        2077 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      object             
 9   expanded_urls               2077 non-null   object             
 10  rating_numerator            2077 non-null   int64           

##### Define

8. Drop the values of rating numerator that is greater or equal to 15

##### Code

In [269]:
df_1_clean = df_1_clean[df_1_clean.rating_numerator < 15]

##### Test

In [270]:
df_1_clean.rating_numerator.value_counts().sort_index(ascending = False)

14     38
13    287
12    485
11    413
10    434
9     152
8      98
7      51
6      32
5      34
4      15
3      19
2       9
1       4
0       1
Name: rating_numerator, dtype: int64

In [271]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2072 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2072 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2072 non-null   datetime64[ns, UTC]
 4   source                      2072 non-null   object             
 5   text                        2072 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      object             
 9   expanded_urls               2072 non-null   object             
 10  rating_numerator            2072 non-null   int64           

> DATA TIDINESS issue:

##### Define

There are 4 columns for dog stages: doggo, floofer, pupper, puppo.

Melt the 4 columns with the dog stages:

- Make a transitory columns called "none" to store the None values; utilize lowercase 'n' to stay away from disarray with 'None' special NA values.
- Make another column stage to store the classifications: doggo, floofer, pupper, puppo, just as None, and select the stage from the segment that contains a values.
- Cast the stage column to categorical data type. 
- Drop the 4 unique dog stage column, and the transitory "none" column.

##### Code

In [272]:
# I will replace the stage name with 1, and 'None' with 0, like a dummy variable
dummy_data = lambda x: 0 if x == 'None' else 1

df_1_clean.doggo = df_1_clean.doggo.apply(dummy_data)
df_1_clean.floofer = df_1_clean.floofer.apply(dummy_data)
df_1_clean.pupper = df_1_clean.pupper.apply(dummy_data)
df_1_clean.puppo = df_1_clean.puppo.apply(dummy_data)

# to see the count for none column and set stages by adding 
df_1_clean['none'] = df_1_clean['doggo'] + df_1_clean['floofer'] + \
                     df_1_clean['pupper'] + df_1_clean['puppo']

# investigate the none value column
df_1_clean['none'].value_counts()

0    1738
1     323
2      11
Name: none, dtype: int64

The number 2 meant that we have 11 tweets that have 2 dog staged set which contain multiple dogs

##### Test

In [273]:
#Algorithm sources: https://code.i-harness.com/en/q/19c9fbc 

# if there are NO stages specified then set 'None' to 1
stage_none = lambda x: 1 if x == 0 else 0

# reset values in 'none' 
df_1_clean['none'] = df_1_clean['none'].apply(stage_none)

# Order the stages in increasing count order: floofer, puppo, doggo and pupper
# set the choice order for dog stage based on count order
stage = ['floofer', 'puppo', 'doggo', 'pupper', 'none']

# set the conditions for selecting the dog stage based on count order
conditions = [
    (df_1_clean[stage[0]] == 1),
    (df_1_clean[stage[1]] == 1),
    (df_1_clean[stage[2]] == 1),
    (df_1_clean[stage[3]] == 1),
    (df_1_clean[stage[4]] == 1)]

# select the dog stage based on the first successful condition; stage[4] is 'None'
df_1_clean['stage'] = np.select(conditions, stage, default = stage[4])

# now we can drop the original 4 dog stage columns, AND the temporary 'None'
df_1_clean.drop(stage, axis = 1, inplace = True)

# set the 'stage' column data type to category
df_1_clean['stage'] = df_1_clean.stage.astype('category')

In [274]:
df_1_clean.stage.value_counts()

none       1738
pupper      219
doggo        81
puppo        24
floofer      10
Name: stage, dtype: int64

In [275]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2072 entries, 0 to 2355
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2072 non-null   int64              
 1   in_reply_to_status_id       0 non-null      float64            
 2   in_reply_to_user_id         0 non-null      float64            
 3   timestamp                   2072 non-null   datetime64[ns, UTC]
 4   source                      2072 non-null   object             
 5   text                        2072 non-null   object             
 6   retweeted_status_id         0 non-null      float64            
 7   retweeted_status_user_id    0 non-null      float64            
 8   retweeted_status_timestamp  0 non-null      object             
 9   expanded_urls               2072 non-null   object             
 10  rating_numerator            2072 non-null   int64           

In [276]:
#drop columns with zero values 
df_1_clean = df_1_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id',
                              'retweeted_status_id','retweeted_status_user_id',
                              'retweeted_status_timestamp'], axis = 1) 

In [277]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2072 entries, 0 to 2355
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2072 non-null   int64              
 1   timestamp           2072 non-null   datetime64[ns, UTC]
 2   source              2072 non-null   object             
 3   text                2072 non-null   object             
 4   expanded_urls       2072 non-null   object             
 5   rating_numerator    2072 non-null   int64              
 6   rating_denominator  2072 non-null   int64              
 7   name                2072 non-null   object             
 8   stage               2072 non-null   category           
dtypes: category(1), datetime64[ns, UTC](1), int64(3), object(4)
memory usage: 147.9+ KB


In [278]:
df_1_clean.head()

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name stage  
0                  10   Phineas  none  
1                  10     Tilly  none  
2                  10    Archie  none  
3                  10     Darla  none  
4                  10  Franklin  none

<strong>Tweet Image Perdiction</strong> aka. <strong>df_2 data</strong>

> DATA TIDINESS ISSUE: 

##### Define

For this table there are multiple values in the three prediction that the algorithm failed to predict. I will first create a column in the prediction table then I will merge with the archive table. 

1. I will create two new columns in prediction called "breed" and "confidence" which will check each dog breed prediction flag in order (p1_dog, p2_dog, p3_dog) and copy the breed with the highest confidence level into breed column. 
2. copy the associate confidence level into new confidence column
3. For those breed that are FALSE in value I will set to none for breed and confidence to 0
4. I will merge the breed and confidence column into the archive table
5. Turn confidence column into percentage value 

sources use for this code: https://code.i-harness.com/en/q/19c9fbc

##### Code

In [279]:
# setup the conditions for selection
conditions = [(df_2_clean['p1_dog'] == True),
              (df_2_clean['p2_dog'] == True),
              (df_2_clean['p3_dog'] == True)]

# set the choice order based on the selection conditions for predicted breed
choices_breed = [df_2_clean['p1'], 
                 df_2_clean['p2'],
                 df_2_clean['p3']]

# set the choice order for confidence level based on the selection conditions
choices_confidence = [df_2_clean['p1_conf'], 
                      df_2_clean['p2_conf'], 
                      df_2_clean['p3_conf']]

# select the predicted breed based on the first successful condition
df_2_clean['breed'] = np.select(conditions, choices_breed, default = 'none')

# select the predicted confidence level based on the first successful condition
df_2_clean['confidence'] = np.select(conditions, choices_confidence, default = 0)

In [280]:
# merge the breed and confidence columns to archive
mask_cols = ['tweet_id', 'breed', 'confidence']
df_1_clean = pd.merge(df_1_clean, df_2_clean[mask_cols], on = 'tweet_id', how = 'inner')

# change data type for the confidence interval to a rounded, whole integer (%)
df_1_clean.confidence = (df_1_clean.confidence * 100).astype(int)

# after the two columns are merged, we can drop them from the 'predictions' table
df_2_clean.drop(['breed', 'confidence'], axis = 1, inplace = True)

##### Test

In [281]:
df_2_clean.head(10)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5  666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6  666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7  666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8  666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9  666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   
5        1    Bernese_mountain_dog  0.651137    True    English_springer   
6        1              box_turtle  0.933012   False          mud_turtle   
7        1                    chow  0.692517    True     Tibetan_mastiff   
8        1           shopping_cart  0.962465   False     shopping_basket   
9        1        miniature_poodle  0.201493    True            komondor   

    p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0  0.156665    True            Shetland_sheepdog  0.061428    True  
1  0.074192    True          Rhodesian_ridgeback  0.072010    True  
2  0.138584    True                   bloodhound  0.116197    True  
3  0.360687    True           miniature_pinscher  0.222752    True  
4  0.243682    True                     Doberman  0.154629    True  
5  0.263788    True   Greater_Swiss_Mountain_dog  0.016199    True  
6  0.045885   False                     terrapin  0.017885   False  
7  0.058279    True                     fur_coat  0.054449   False  
8  0.014594   False             golden_retriever  0.007959    True  
9  0.192305    True  soft-coated_wheaten_terrier  0.082086    True

In [282]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1949 entries, 0 to 1948
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            1949 non-null   int64              
 1   timestamp           1949 non-null   datetime64[ns, UTC]
 2   source              1949 non-null   object             
 3   text                1949 non-null   object             
 4   expanded_urls       1949 non-null   object             
 5   rating_numerator    1949 non-null   int64              
 6   rating_denominator  1949 non-null   int64              
 7   name                1949 non-null   object             
 8   stage               1949 non-null   category           
 9   breed               1949 non-null   object             
 10  confidence          1949 non-null   int32              
dtypes: category(1), datetime64[ns, UTC](1), int32(1), int64(3), object(5)
memory usage: 162.0+ KB


In [283]:
df_1_clean.head()

tweet_id                 timestamp              source  \
0  892420643555336193 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1  892177421306343426 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2  891815181378084864 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3  891689557279858688 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4  891327558926688256 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name stage               breed  confidence  
0                  10   Phineas  none                none           0  
1                  10     Tilly  none           Chihuahua          32  
2                  10    Archie  none           Chihuahua          71  
3                  10     Darla  none  Labrador_retriever          16  
4                  10  Franklin  none              basset          55

<strong>Json Tweet Data</strong> aka. <strong>df_3 data</strong>

> DATA TIDINESS ISSUE: 

##### Define

The Json Tweet Data should be merge with archive data. I will merge the retweet_count and favorite_count columns into the archive table. First I rename the id to tweet_id and will drop every column in the JSON table except Tweed Id, retweet_count and favorite_count.

##### Code

In [284]:
df_3_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2354 non-null   datetime64[ns, UTC]
 1   id                             2354 non-null   int64              
 2   id_str                         2354 non-null   int64              
 3   full_text                      2354 non-null   object             
 4   truncated                      2354 non-null   bool               
 5   display_text_range             2354 non-null   object             
 6   entities                       2354 non-null   object             
 7   extended_entities              2073 non-null   object             
 8   source                         2354 non-null   object             
 9   in_reply_to_status_id          78 non-null     float64            
 10  in_reply_to_status_id_st

In [285]:
df_3_clean.head()

created_at                  id              id_str  \
0 2017-08-01 16:23:56+00:00  892420643555336193  892420643555336192   
1 2017-08-01 00:17:27+00:00  892177421306343426  892177421306343424   
2 2017-07-31 00:18:03+00:00  891815181378084864  891815181378084864   
3 2017-07-30 15:58:51+00:00  891689557279858688  891689557279858688   
4 2017-07-29 16:00:24+00:00  891327558926688256  891327558926688256   

                                           full_text  truncated  \
0  This is Phineas. He's a mystical boy. Only eve...      False   
1  This is Tilly. She's just checking pup on you....      False   
2  This is Archie. He is a rare Norwegian Pouncin...      False   
3  This is Darla. She commenced a snooze mid meal...      False   
4  This is Franklin. He would like you to stop ca...      False   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1           [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2           [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3            [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4           [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   
1  {'media': [{'id': 892177413194625024, 'id_str'...   
2  {'media': [{'id': 891815175371796480, 'id_str'...   
3  {'media': [{'id': 891689552724799489, 'id_str'...   
4  {'media': [{'id': 891327551943041024, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   
1  <a href="http://twitter.com/download/iphone" r...                    NaN   
2  <a href="http://twitter.com/download/iphone" r...                    NaN   
3  <a href="http://twitter.com/download/iphone" r...                    NaN   
4  <a href="http://twitter.com/download/iphone" r...                    NaN   

   ...  favorite_count  favorited  retweeted possibly_sensitive  \
0  ...           39467      False      False                0.0   
1  ...           33819      False      False                0.0   
2  ...           25461      False      False                0.0   
3  ...           42908      False      False                0.0   
4  ...           41048      False      False                0.0   

  possibly_sensitive_appealable  lang  retweeted_status quoted_status_id  \
0                           0.0    en               NaN              NaN   
1                           0.0    en               NaN              NaN   
2                           0.0    en               NaN              NaN   
3                           0.0    en               NaN              NaN   
4                           0.0    en               NaN              NaN   

   quoted_status_id_str  quoted_status  
0                   NaN            NaN  
1                   NaN            NaN  
2                   NaN            NaN  
3                   NaN            NaN  
4                   NaN            NaN  

[5 rows x 31 columns]

In [286]:
#rename id to tweet_id 
df_3_clean.rename(columns = {'id':'tweet_id'}, inplace = True) 

In [288]:
#drop every column except tweet_id, retweet_count and favorite_count
df_3_clean.drop(df_3_clean.columns.difference(['tweet_id','retweet_count','favorite_count']), 1, inplace=True)

In [290]:
#merge with archive table
df_1_clean = pd.merge(df_1_clean, df_3_clean,on = 'tweet_id', how = 'left')

##### Test

In [291]:
df_3_clean.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

In [293]:
df_1_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1949 entries, 0 to 1948
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            1949 non-null   int64              
 1   timestamp           1949 non-null   datetime64[ns, UTC]
 2   source              1949 non-null   object             
 3   text                1949 non-null   object             
 4   expanded_urls       1949 non-null   object             
 5   rating_numerator    1949 non-null   int64              
 6   rating_denominator  1949 non-null   int64              
 7   name                1949 non-null   object             
 8   stage               1949 non-null   category           
 9   breed               1949 non-null   object             
 10  confidence          1949 non-null   int32              
 11  retweet_count       1949 non-null   int64              
 12  favorite_count      1949 non-null 

In [296]:
df_1_clean.isnull().sum()

tweet_id              0
timestamp             0
source                0
text                  0
expanded_urls         0
rating_numerator      0
rating_denominator    0
name                  0
stage                 0
breed                 0
confidence            0
retweet_count         0
favorite_count        0
dtype: int64

<a id='sd'></a>
### Store Data 
<a href="#top">Back to Table of Content</a>

This is the end of the clean data phase where the data_1_clean aka. Tweet Archive Enhanced has been tidy and the quality of the data is much better than before. I will store the data into "twitter_archive_master.csv" file

In [297]:
df_1_clean.to_csv('twitter_archive_master.csv', index = False)